### ISSUE TITLE
Allow implementing traits with statically uncallable methods

### ISSUE THREAD

**Issue by Stebalien** [rust-lang/rust#35369]
Basically, given:

```rust
trait Foo {
    fn foo(self) where Self: Sized;
    fn bar(&self);
}
```

The following is valid:

```rust
impl<T: ?Sized> Foo for T {
    fn foo(self) where Self: Sized {}
    fn bar(&self) {}
}
```

But the following isn't:

```rust
impl<T> Foo for [T] {
    fn foo(self) where Self: Sized { unreachable!() } // Should disappear at compile-time.
    fn bar(&self) {}
}
```

This means that, for example, `Iterator` can't be implemented types that are known to be DSTs, only types that *may* be DSTs. This is an inconsistent and an unnecessary limitation.

**Comment by nikomatsakis** [comment]
Dup of https://github.com/rust-lang/rust/issues/20021

**Comment by nikomatsakis** [comment]
(I closed in favor of #20021 as it has more discussion)
